### PR TITLE
Fix parsing of query with nested object

### PR DESCRIPTION
**Pull request by colincasey**
This modifies the parsing routine of data api queries to correctly parse result sets that include nested `QueriedRecord` data.

[W-12157953](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001FiOtaYAF/view)

### CODE REVIEW

File: salesforce_functions/data_api/_requests.py [hunk 1]
```diff
@@ -229,42 +229,47 @@ async def _process_records_response(
...
+                sub_query_results[key] = await _parse_record_query_result(value, download_file_fn)
+        elif _is_binary_field(salesforce_object_type, key):
+            fields[key] = await download_file_fn(value)
+        else:
+            fields[key] = value
+    return QueriedRecord(salesforce_object_type, fields, sub_query_results)
+
+
 def _is_binary_field(salesforce_object_type: str, field_name: str) -> bool:
     return salesforce_object_type == "ContentVersion" and field_name == "VersionData"
```
Comment (edmorley) [inline salesforce_functions/data_api/_requests.py:272]: (flake8 is failing with this newline removed)

```suggestion


```

File: salesforce_functions/data_api/_requests.py [hunk 2]
```diff
@@ -229,42 +229,47 @@ async def _process_records_response(
...
+        if isinstance(value, dict):
+            if "attributes" in value:
+                fields[key] = await _parse_queried_record(value, download_file_fn)
+            else:
+                sub_query_results[key] = await _parse_record_query_result(value, download_file_fn)
+        elif _is_binary_field(salesforce_object_type, key):
+            fields[key] = await download_file_fn(value)
+        else:
+            fields[key] = value
+    return QueriedRecord(salesforce_object_type, fields, sub_query_results)
```
Comment (edmorley) [inline salesforce_functions/data_api/_requests.py:267]:
```suggestion
            fields[key] = value

    return QueriedRecord(salesforce_object_type, fields, sub_query_results)
```

File: tests/test_data_api.py [hunk 3]
```diff
@@ -645,3 +645,29 @@ async def test_unit_of_work_single_delete() -> None:
     assert result == {
         ReferenceId("referenceId0"): "a01B0000009gSr9IAE",
     }
+
```
Comment (edmorley) [inline tests/test_data_api.py:648]:
```suggestion


```

**Review by edmorley — APPROVED** [review]
Thank you!

File: salesforce_functions/data_api/_requests.py [hunk 4]
```diff
@@ -229,38 +229,45 @@ async def _process_records_response(
...
-                QueriedRecord(salesforce_object_type, fields, sub_query_results)
-            )
-
-        return RecordQueryResult(done, total_size, records, next_records_url)
+        return await _parse_record_query_result(json_body, download_file_fn)
 
     raise UnexpectedRestApiResponsePayload()
 
 
+async def _parse_record_query_result(json_body: dict, download_file_fn: DownloadFileFunction) -> RecordQueryResult:
```
Comment (edmorley) [inline salesforce_functions/data_api/_requests.py:237]: Mypy is failing with:

```
salesforce_functions/data_api/_requests.py:237: error: Missing type parameters
for generic type "dict"  [type-arg]
    async def _parse_record_query_result(json_body: dict, download_file_fn...
                                                    ^
salesforce_functions/data_api/_requests.py:249: error: Missing type parameters
for generic type "dict"  [type-arg]
    async def _parse_queried_record(record_json: dict, download_file_fn: D...
                                                 ^
salesforce_functions/data_api/_requests.py:264: error: Incompatible types in
assignment (expression has type "bytes", target has type "QueriedRecord") 
[assignment]
                fields[key] = await download_file_fn(value)
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (colincasey) [reply salesforce_functions/data_api/_requests.py]: i figured out how to run `pylint`.  didn't realize `mypy` was another check

**Pull request merged**